File: backend/views/settings_routes.py
from fastapi import APIRouter, Depends, HTTPException, Request
from typing import Dict, Any
from pydantic import BaseModel, Field

from controllers.settings_controller import SettingsController
from models.user import User
from dependencies import get_current_user

router = APIRouter(prefix="/api/settings", tags=["settings"])
public_router = APIRouter(prefix="/api/settings", tags=["settings"])

class SettingsUpdate(BaseModel):
    work_hours_per_day: int = Field(ge=1, le=24, description="Work hours per day (1-24)")
    work_days_per_month: int = Field(ge=1, le=31, description="Work days per month (1-31)")
    months_in_year: int = Field(ge=1, le=12, description="Months in year (1-12)")

class PasswordVerify(BaseModel):
    password: str = Field(min_length=1, description="Site access password")

class PasswordUpdate(BaseModel):
    current_password: str = Field(min_length=1, description="Current site password")
    new_password: str = Field(min_length=1, description="New site password")

@router.get("", response_model=Dict[str, Any])
async def get_settings(current_user: User = Depends(get_current_user)):
    """Get current business rules settings - accessible to all roles"""
    return SettingsController.get_settings()

@router.put("", response_model=Dict[str, Any])
async def update_settings(
    settings: SettingsUpdate,
    current_user: User = Depends(get_current_user)
):
    """Update business rules settings - accessible to all roles"""
    return SettingsController.update_settings(settings.model_dump())

@public_router.post("/verify-password")
async def verify_site_password(data: PasswordVerify):
    """Verify site access password - public endpoint, returns signed token"""
    if SettingsController.verify_site_password(data.password):
        token = SettingsController.generate_site_token()
        return {"success": True, "message": "Access granted", "token": token}
    raise HTTPException(status_code=401, detail="Invalid password")

@public_router.post("/verify-token")
async def verify_site_token(request: Request):
    """Verify a site access token is still valid"""
    from fastapi import Request as _R
    token = request.headers.get("X-Site-Token", "")
    if token and SettingsController.verify_site_token(token):
        return {"valid": True}
    raise HTTPException(status_code=401, detail="Invalid or expired token")

@router.put("/site-password")
async def update_site_password(
    data: PasswordUpdate,
    current_user: User = Depends(get_current_user)
):
    """Update site access password - requires authentication"""
    if not SettingsController.verify_site_password(data.current_password):
        raise HTTPException(status_code=400, detail="Current password is incorrect")
    
    try:
        SettingsController.update_site_password(data.new_password)
        return {"success": True, "message": "Site password updated successfully"}
    except Exception as e:
        raise HTTPException(status_code=500, detail=str(e))

File: backend-node/src/routes/settingsRoutes.js
const express = require('express');
const router = express.Router();
const SettingsController = require('../controllers/settingsController');
const User = require('../models/user');
const { authenticate } = require('../middleware/auth');
const db = require('../database');

// Get settings - requires authentication
router.get('/', authenticate, (req, res) => {
  try {
    res.json(SettingsController.getSettings());
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

// Update settings - requires authentication
router.put('/', authenticate, (req, res) => {
  try {
    const { work_hours_per_day, work_days_per_month, months_in_year } = req.body;

    // Validate
    if (work_hours_per_day !== undefined && (work_hours_per_day < 1 || work_hours_per_day > 24)) {
      return res.status(400).json({ detail: 'Work hours per day must be between 1 and 24' });
    }
    if (work_days_per_month !== undefined && (work_days_per_month < 1 || work_days_per_month > 31)) {
      return res.status(400).json({ detail: 'Work days per month must be between 1 and 31' });
    }
    if (months_in_year !== undefined && (months_in_year < 1 || months_in_year > 12)) {
      return res.status(400).json({ detail: 'Months in year must be between 1 and 12' });
    }

    const result = SettingsController.updateSettings(req.body);
    res.json(result);
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

// Verify site password - public endpoint
router.post('/verify-password', (req, res) => {
  try {
    const { password } = req.body;

    if (!password) {
      return res.status(400).json({ detail: 'Password is required' });
    }

    if (SettingsController.verifySitePassword(password)) {
      res.json({ success: true, message: 'Access granted' });
    } else {
      res.status(401).json({ detail: 'Invalid password' });
    }
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

// Update site password - requires authentication
router.put('/site-password', authenticate, (req, res) => {
  try {
    const { current_password, new_password } = req.body;

    if (!current_password || !new_password) {
      return res.status(400).json({ detail: 'Current password and new password are required' });
    }

    if (!SettingsController.verifySitePassword(current_password)) {
      return res.status(400).json({ detail: 'Current password is incorrect' });
    }

    SettingsController.updateSitePassword(new_password);
    res.json({ success: true, message: 'Site password updated successfully' });
  } catch (e) {
    res.status(500).json({ detail: e.message });
  }
});

// Change user password - requires authentication
router.put('/change-password', authenticate, (req, res) => {
  try {
    const { current_password, new_password } = req.body;

    if (!current_password || !new_password) {
      return res.status(400).json({ detail: 'Current password and new password are required' });
    }

    if (new_password.length < 4) {
      return res.status(400).json({ detail: 'New password must be at least 4 characters' });
    }

    // Verify current password
    if (!req.currentUser.verifyPassword(current_password)) {
      return res.status(400).json({ detail: 'Current password is incorrect' });
    }

    // Update password
    const newHash = User.hashPassword(new_password);
    db.execute(
      'UPDATE users SET password_hash = ?, updated_at = CURRENT_TIMESTAMP WHERE id = ?',
      [newHash, req.currentUser.id]
    );

    res.json({ success: true, message: 'Password updated successfully' });
  } catch (e) {
    res.status(500).json({ detail: `Failed to update password: ${e.message}` });
  }
});

module.exports = router;
